client/authn.go

```diff
@@ -0,0 +1,199 @@
+// itself by calling registerAuthResponder.
+type authResponder interface {
+	// Scheme should return the name of the authorization scheme for which
+	// the responder should be called.
+	scheme() string
+	// authRespond, given the authentication header value associated with
+	// the scheme that it implements, can decide if the request should be
+	// retried.  If it returns true, then the request is retransmitted to
+	// the server, presumably because it has added an authentication header
+	// which it believes the server will accept.
+	authRespond(cli *Client, challenge string, req *http.Request) (bool, error)
+	// AuthCompleted, given a (possibly empty) WWW-Authenticate header and
+	// a successful response, should decide if the server's reply should be
+	// accepted.
+	authCompleted(cli *Client, challenge string, resp *http.Response) (bool, error)
+}
+
+// authResponderCreator either creates a new authResponder, or returns nil
+type authResponderCreator func() authResponder
+
+var authResponderCreators = []authResponderCreator{}
+
+// registerAuthResponder registers a function which will be called at startup
+// to create an authResponder.
+func registerAuthResponder(arc authResponderCreator) {
+	authResponderCreators = append(authResponderCreators, arc)
+}
+
+// Run through all of the registered responder creators and build a map of
+// names-to-responders.
+func createAuthResponders() map[string]authResponder {
+	ars := make(map[string]authResponder)
+	for _, arc := range authResponderCreators {
+		responder := arc()
+		if responder != nil {
+			scheme := strings.ToLower(responder.scheme())
+			ars[scheme] = responder
+		}
+	}
+	return ars
+}
+
+// doer is just an interface wrapper that we use to be able to Do() using
+// either an http.Client or an httputil.Clientconn.
+type doer interface {
+	Do(req *http.Request) (resp *http.Response, err error)
+}
+
+// Run the Doer's Do() method, setting and saving cookies.  An http.Client
+// would handle this for us if we passed in a cookie jar when we initialized
+// it, but we'd still have to do the work for httputil.ClientConn, so we don't
+// set it for http.Clients to avoid duplicating cookies in requests.
+func (cli *Client) doWithCookies(doer doer, req *http.Request) (*http.Response, error) {
+	if localhost, err := os.Hostname(); err == nil && cli.proto == "unix" {
+		req.URL.Host = localhost
+	} else {
+		req.URL.Host = cli.addr
+	}
+	req.URL.Scheme = cli.scheme
+	if cli.jar != nil {
+		for _, cookie := range cli.jar.Cookies(req.URL) {
+			req.AddCookie(cookie)
+		}
+	}
+	resp, err := doer.Do(req)
+	if cli.jar != nil && resp != nil {
+		if cookies := resp.Cookies(); cookies != nil {
+			cli.jar.SetCookies(req.URL, cookies)
+		}
+	}
+	return resp, err
+}
+
+// doWithAuthn calls Do(), and handles any "unauthorized" errors
+// which it returns by retrying it with authentication.
+func (cli *Client) doWithAuthn(doer doer, req *http.Request, body []byte) (*http.Response, error) {
```

Comment:
Hmm, Reader isn't going to be good enough, since if we get a 401 error and have to reattempt a request with authentication, we'll need to be able to supply the same POST data that we sent the first time.
We'd been reading it all into a byte array, first thing, and just referring to it every time, but that can get expensive.
I think we can add a rewindable wrapper around the body Reader, one that hangs on to everything that's read through it, so that it can be explicitly rewound before every attempt at hitting the server.  If the server comes back with an error fast enough, maybe the client won't have run through too much of the input data.